beets/util/artresizer.py

```diff
@@ -83,14 +84,50 @@ def pil_resize(maxwidth, path_in, path_out=None, quality=0):
             quality = -1
 
         im.save(util.py3_path(path_out), quality=quality)
-        return path_out
+        if max_filesize > 0:
+            # If maximum filesize is set, we attempt to lower the quality of
+            # jpeg conversion by a proportional amount, up to 3 attempts
+            # First, set the maximum quality to either provided, or 95
+            if quality > 0:
+                lower_qual = quality
+            else:
+                lower_qual = 95
+            for i in range(3):
+                # 3 attempts is an abitrary choice
+                filesize = os.stat(path_out).st_size
+                log.debug(u"PIL Pass {0} : Output size: {1}B", i, filesize)
+                if filesize <= max_filesize:
+                    return path_out
+                # Calculate percentage decrease required. This is approximate,
+                # as quality may only loosely correlate with filesize.
+                frac_diff = 1 - ((filesize - max_filesize) / max_filesize)
+                assert frac_diff < 1
+                lower_qual = lower_qual * frac_diff
+                # Restrict quality dropping below 10
+                if lower_qual < 10:
+                    lower_qual = 10
+                # Use optimize flag to improve filesize decrease
+                im.save(
+                    util.py3_path(path_out), quality=lower_qual, optimize=True
+                )
+            log.error(u"PIL Failed to resize file to below {0}B", max_filesize)
+            raise IOError
```

Comment:
Maybe return the resized image here anyway? That's presumably still closer to the target size than the original.

Reply:
Yeah, I've done that now. Should I adjust the log to a warning?

Reply:
That sounds like a good idea!